File: .NET Framework 4 Calendar Week Range Utility/[C#][VB]-.NET Framework 4 Calendar Week Range Utility/C# and VB/C#,VB/WeekRangeCS/Utility.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.Text.RegularExpressions;

public enum JapaneseEras 
{
   Heisei = 4,
   Showa = 3,
   Taisho = 2,
   Meiji = 1,
}

public class CalendarComparer : IComparer
{
   public int Compare(object x, object y)
   {
      Type tX = x as Type;
      Type tY = y as Type;

      return tX.Name.CompareTo(tY.Name);
   }
}

internal static class Utility
{
   internal static Dictionary<string, DateTime> StartDates;

   static Utility() {
      StartDates = new Dictionary<string, DateTime>();
      StartDates.Add("Hebrew", new DateTime(1583, 9, 17));
   }

   internal static bool IsValidCalendar(Type t)
   {

      if (t.ToString().EndsWith("Calendar"))
         if (t.IsAbstract)
            return false;
         else
            return true;
      else
         return false;
   }

   internal static string GetProperName(Type calendarType)
   {
      return Regex.Replace(calendarType.Name, @"(\p{Lu}\p{Ll}+)(\p{Lu}\p{Ll}+)*(Calendar)", AddSpaces);
   }

   internal static string AddSpaces(Match m)
   {
      if (m.Groups[2].Captures.Count <= 1)
      {
         return m.Groups[1] + " " + m.Groups[2];
      }
      else
      {
         string name = m.Groups[1].Value;
         foreach (Capture capture in m.Groups[2].Captures)
            name += " " + capture.Value;
         return name;
      }
   }

   internal static string GetNonDefaultCalendarDate(DateTime date, Calendar cal)
   {
      return String.Format(@"{0:d2}/{1:d2}/{2:d4}", cal.GetMonth(date), cal.GetDayOfMonth(date), cal.GetYear(date));
   }
}

File: .NET Framework 4 Calendar Week Range Utility/[C#][VB]-.NET Framework 4 Calendar Week Range Utility/C# and VB/C#,VB/WeekRangeVB/Utility.vb
Imports System.Collections
Imports System.Collections.Generic
Imports System.Globalization
Imports System.Text.RegularExpressions

Public Enum JapaneseEras As Integer
   Heisei = 4
   Showa = 3
   Taisho = 2
   Meiji = 1
End Enum

Public Class CalendarComparer : Implements IComparer
   Public Function Compare(ByVal x As Object, ByVal y As Object) As Integer _
                  Implements IComparer.Compare
      Dim tX As Type = DirectCast(x, Type)
      Dim tY As Type = DirectCast(y, Type)

      Return tX.Name.CompareTo(tY.Name)
   End Function
End Class

Module Utility
    Friend StartDates As Dictionary(Of String, DateTime)

    Sub New()
        StartDates = New Dictionary(Of String, DateTime)()
        StartDates.Add("Hebrew", New DateTime(1583, 9, 17))
    End Sub

    Friend Function IsValidCalendar(ByVal t As Type) As Boolean
        If t.ToString().EndsWith("Calendar") Then
            If t.IsAbstract Then
                Return False
            Else
                Return True
            End If
        Else
            Return False
        End If
    End Function

    Friend Function GetProperName(ByVal calendarType As Type) As String
        Return Regex.Replace(calendarType.Name, "(\p{Lu}\p{Ll}+)(\p{Lu}\p{Ll}+)*(Calendar)", AddressOf AddSpaces)
    End Function

    Friend Function AddSpaces(ByVal m As Match) As String
        If m.Groups(2).Captures.Count <= 1 Then
            Return m.Groups(1).Value + " " + m.Groups(2).Value
        Else
            Dim name As String = m.Groups(1).Value
            For Each capture As Capture In m.Groups(2).Captures
                name += " " + capture.Value
            Next
            Return name
        End If
    End Function

    Friend Function GetNonDefaultCalendarDate(ByVal dat As DateTime, ByVal cal As Calendar) As String
        Return String.Format("{0:d2}/{1:d2}/{2:d4}", cal.GetMonth(dat), cal.GetDayOfMonth(dat), cal.GetYear(dat))
    End Function
End Module
